src/System.Private.CoreLib/src/System/Runtime/Loader/ComponentDependencyResolver.cs

```diff
@@ -0,0 +1,282 @@
+#endif
+
+        private IEnumerable<LibraryNameVariation> DetermineLibraryNameVariations(string libName, bool isRelativePath)
+        {
+            // This is a copy of the logic in DetermineLibNameVariations in dllimport.cpp in CoreCLR
+
+            if (!isRelativePath)
+            {
+                yield return new LibraryNameVariation(string.Empty, string.Empty);
+            }
+            else
+            {
+                bool containsSuffix = false;
+                int indexOfSuffix = libName.IndexOf(LibraryNameSuffix);
+                if (indexOfSuffix >= 0)
+                {
+                    indexOfSuffix += LibraryNameSuffix.Length;
+                    containsSuffix = indexOfSuffix == libName.Length || libName[indexOfSuffix] == '.';
+                }
+
+                bool containsDelim = libName.Contains(Path.DirectorySeparatorChar);
+
+                if (containsSuffix)
+                {
+                    yield return new LibraryNameVariation(string.Empty, string.Empty);
+                    if (!containsDelim)
+                    {
+                        yield return new LibraryNameVariation(LibraryNamePrefix, string.Empty);
+                    }
+                    yield return new LibraryNameVariation(string.Empty, LibraryNameSuffix);
+                    if (!containsDelim)
+                    {
+                        yield return new LibraryNameVariation(LibraryNamePrefix, LibraryNameSuffix);
+                    }
+                }
+                else
+                {
+                    yield return new LibraryNameVariation(string.Empty, LibraryNameSuffix);
+                    if (!containsDelim)
+                    {
+                        yield return new LibraryNameVariation(LibraryNamePrefix, LibraryNameSuffix);
+                    }
+                    yield return new LibraryNameVariation(string.Empty, string.Empty);
+                    if (!containsDelim)
+                    {
+                        yield return new LibraryNameVariation(LibraryNamePrefix, string.Empty);
+                    }
+                }
+            }
+
+            yield return new LibraryNameVariation(string.Empty, string.Empty);
+
+            if (isRelativePath &&
+                !libName.EndsWith(".dll", StringComparison.OrdinalIgnoreCase) &&
+                !libName.EndsWith(".exe", StringComparison.OrdinalIgnoreCase))
+            {
+                yield return new LibraryNameVariation(string.Empty, LibraryNameSuffix);
+            }
+        }
+#endif
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl, CharSet = HostpolicyCharSet)]
+        internal delegate void corehost_resolve_component_dependencies_result_fn(
+            string assembly_paths,
+            string native_search_paths,
+            string resource_search_paths);
+
+#pragma warning disable BCL0015 // Disable Pinvoke analyzer errors.
+        [DllImport("hostpolicy", CharSet = HostpolicyCharSet)]
```

Comment:
Are we sure we want to take this dependency in `CoreLib` on `hostpolicy`? Also, it feels strange to have a `corefx` API depend on the `hostpolicy` assembly. What will happen in other scenarios that aren't using the `dotnet.exe` host? For example, Mono/Xamarin.

/cc @ericstj @jkotas

Reply:
The actual behavior of the API is very runtime specific. On .NET Core it relies on `.deps.json`. On .NET Framework (if we would have it there) it would rely on something completely different (probably just directory layout). On Mono I assume it will also rely on different mechanism, whatever is it that Mono uses to describe dependencies.

The dependency on `hostpolicy` is for two reasons:
* Right now to avoid code duplication - all the logic around `.deps.json` is in `hostpolicy` and duplicating it elsewhere is just asking for bugs.
* Technological limitations - we don't have a JSON parser which could be part of CoreLib. The eventual long term plan is to use the new JSON parser from ASP.NET and move the code from `hostpolicy` to CoreLib for both startup `.deps.json` processing as well as for this functionality.

As for what happens if the runtime is not hosted through `hostpolicy`, currently this API would fail. I'm still thinking of what is doable for .NET Core 3 RTM, options are:
* Fallback to directory based behavior (basically just like if there's no `.deps.json`)
* Introduce a mechanism for custom coreclr hosts to provide the same service as `hostpolicy`. Currently they basically have to do that for startup, so it would make sense to have that.

Reply:
Would it make more sense to add these APIs to the [`Microsoft.Extensions.DependencyModel`](https://github.com/dotnet/core-setup/tree/master/src/managed/Microsoft.Extensions.DependencyModel) library instead? This library already knows how to read and interpret `.deps.json` files.

Reply:
The problem is layering:
* That library has dependency on `Newtonsoft.Json` which is problematic (dependency on a specific version, which the app using the new API may not like, and so on)
* Going forward we think we will need this API inside CoreLib itself. The main scenario is COM activation where CoreLib would use this API to be able to load the necessary dependencies for the assembly which contains the COM object being activated.

Reply:
> That library has dependency on Newtonsoft.Json which is problematic (dependency on a specific version, which the app using the new API may not like, and so on)

I was told that @steveharter was going to convert it to use the new JSON APIs instead on netcoreapp3.0+ (or netstandard2.1, if they make it into that). So that dependency will be going away in newer platforms. See https://github.com/dotnet/core-setup/issues/3311.

> Going forward we think we will need this API inside CoreLib itself.

Ok, that is a good enough reason for me.

Reply:
Thanks for the link to the issue - wasn't aware of it. Steve will not be the one doing it, that would be on me or the other Steve (@sdmaclea). I personally would very much prefer to move the code from `hostpolicy` which is currently all in C++ into corlib. That would solve several issues:
* Our own productivity - C# is way faster to write/maintain than C++
* No dependency on host (whatever that means)
* Easy sharing of that functionality with other pieces - like this one
* Ability to provide consistent diagnostics experience. Today one of the big problems of the host is that it has no ties to EventPipe/ETW, so diagnosing issues which span host and runtime is really tricky. (Adding support for ETW into host is a whole other topic, since it has its own PAL and everything...)

Reply:
Another thing to consider, if you can't move the code down, is to have CoreCLR expose this as part of its hosting APIs and and have hostpolicy implement that hosting extension.  That way you allow for any CoreCLR host to plug into this.  dotnet/hostpolicy isn't the only coreclr host.